### PR TITLE
Self-regenerating zombies

### PR DESCRIPTION
**Pull request by dontbetank**
## About the PR 
Right now getting to crit state as zombie means you have to ghost. That's boring. So I decided to add regen to zombie entity while they lie in critical state or close to it. That should encourage people to actually kill zombie players and spend some time to check if they are dead forever or not.
P.S. Should I instead inject some omnizine inside them? Because I didn't find a way to initialize DamageSpecifier in code normally.
**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: Prentor
- tweak: zombies now regenerate while low on hp

### CODE REVIEW

**Comment by dontbetank** [comment]
Ah man. Github!

File: Content.Server/Zombies/ZombieSystem.cs [hunk 1]
```diff
@@ -156,24 +159,102 @@ private void DoGroan(EntityUid uid, ActiveZombieComponent component)
...
+                if (deadStateData == null)
+                {
+                    return;
+                }
+
+                nextThreshold = deadStateData.Value.threshold;
+            }
+
+
+            // Healing to 80% of HP. Not too big and not too small.
```
Comment (Peptide90) [inline Content.Server/Zombies/ZombieSystem.cs:218]: Personally healing to near full seems too much for existing. 

They should heal enough to get out of crit and perhaps be encouraged to eat the deceased to get above 50%. Although iirc all zombie kills are conversions which is another issue.

Comment (dontbetank) [reply Content.Server/Zombies/ZombieSystem.cs:218]: My bad commenting. They actually heal to have 80% of damage required to drop critical.

Comment (dontbetank) [reply Content.Server/Zombies/ZombieSystem.cs]: My bad commenting. They actually heal to 80% of damage required to drop critical.

File: Content.Server/Zombies/ZombieSystem.cs [hunk 2]
```diff
@@ -156,24 +159,102 @@ private void DoGroan(EntityUid uid, ActiveZombieComponent component)
...
+            component.HealingDamageSpecifier.DamageDict["Pierce"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Blunt"] = -10;
+
+            // Burn
+            component.HealingDamageSpecifier.DamageDict["Heat"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Shock"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Cold"] = -10;
+
+            // Zombies do not receive any Toxic or Airloss damage
+        }
```
Comment (metalgearsloth) [inline Content.Server/Zombies/ZombieSystem.cs:175]: This shouldn't be done on startup and zombies shouldn't be healing 60 damage every 30 seconds.

File: Content.Shared/Zombies/ZombieComponent.cs [hunk 3]
```diff
@@ -60,5 +61,18 @@ public sealed class ZombieComponent : Component
...
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float HealingCooldown = 30;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float HealingAccumulator;
+
+        // Set this up through your disease or something
+        [ViewVariables]
+        public DamageSpecifier HealingDamageSpecifier = new DamageSpecifier();
```
Comment (metalgearsloth) [inline Content.Shared/Zombies/ZombieComponent.cs:76]: These should be datafields and networked like the rest of it.

File: Content.Shared/Zombies/ZombieComponent.cs [hunk 4]
```diff
@@ -60,5 +61,18 @@ public sealed class ZombieComponent : Component
...
         public readonly string ZombieRoleId = "Zombie";
+
+        /// <summary>
+        /// How much time passes between healing attempts
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float HealingCooldown = 30;
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float HealingAccumulator;
```
Comment (metalgearsloth) [inline Content.Shared/Zombies/ZombieComponent.cs:72]: Accumulators are deprecated.

File: Content.Server/Zombies/ZombieSystem.cs [hunk 5]
```diff
@@ -156,24 +159,102 @@ private void DoGroan(EntityUid uid, ActiveZombieComponent component)
...
             base.Update(frameTime);
 
-            foreach (var zombiecomp in EntityQuery<ActiveZombieComponent>())
+            foreach (var zombiecomp in EntityQuery<ZombieComponent>())
             {
-                zombiecomp.Accumulator += frameTime;
-                zombiecomp.LastDamageGroanCooldown -= frameTime;
+                // Only zombies that are alive have this thing
+                if (TryComp<ActiveZombieComponent>(zombiecomp.Owner, out var activeComp))
+                {
```
Comment (metalgearsloth) [inline Content.Server/Zombies/ZombieSystem.cs:235]: You've replaced this with a slower iterator, just use the overload that takes 2 components.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Zombies/ZombieSystem.cs (hunk 2 above)
Comment (dontbetank) [reply Content.Server/Zombies/ZombieSystem.cs:175]: They heal up to 60 damage. Usually they will heal 10-20

File: Content.Server/Zombies/ZombieSystem.cs (hunk 5 above)
Comment (dontbetank) [reply Content.Server/Zombies/ZombieSystem.cs:235]: Zombies in crit get their active component removed so it won't work

Comment (dontbetank) [reply Content.Server/Zombies/ZombieSystem.cs:235]: I found out zombies always have Zombie component, but if they're state is not Alive, active component gets removed. So overload with 2 types won't work properly if zombie is in crit.

**Comment by dontbetank** [comment]
Hell yeah, how to apply regen when surgery comes in?

File: Content.Server/Zombies/ZombieSystem.cs [hunk 6]
```diff
@@ -141,6 +147,14 @@ private void OnMeleeHit(EntityUid uid, ZombieComponent component, MeleeHitEvent
             }
         }
 
+        private void OnInit(EntityUid uid, ZombieComponent? component, MapInitEvent args)
+        {
+            if (!Resolve(uid, ref component))
+                return;
+            if (component.HealingSolution == null)
+                component.HealingSolution = new("Omnizine", 2.5);
+        }
```
Comment (AJCM-git) [inline Content.Server/Zombies/ZombieSystem.cs:156]: dont resolve in a event handler, also why are you doing this on map init?

File: Content.Server/Zombies/ZombieSystem.cs [hunk 7]
```diff
@@ -156,24 +170,105 @@ private void DoGroan(EntityUid uid, ActiveZombieComponent component)
             component.LastDamageGroanCooldown = component.GroanCooldown;
         }
 
+        /// <summary>
+        /// Checks if selected entity is a zombie that requires healing
+        /// </summary>
+        private bool ZombieNeedsHealing(EntityUid ent, ZombieComponent? component)
```
Comment (AJCM-git) [inline Content.Server/Zombies/ZombieSystem.cs:176]: just subscribe to MobStateChanged event

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Resolve conflicts, make zombies only heal once they are in crit and make sure they take at least 2-3 minutes to go from hard crit to minimum-health

File: Content.Server/Zombies/ZombieSystem.cs (hunk 5 above)
Comment (metalgearsloth) [reply Content.Server/Zombies/ZombieSystem.cs:244]: Update the code to not iterate every zombie

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This should also be a separate component.

**Comment by AJCM-git** [comment]
@dontbetank you coming back to this?

**Pull request closed without merging**